### PR TITLE
feat(intent)!: Remove guards from actions.

### PR DESCRIPTION
**Pull request by joachimdekker**
## Description

This PR proposes to remove the guards from action and only use tactics.

## Checklist
- [x] Set the proper pull request name
- [x] Merged main into your branch

### CODE REVIEW

**Review by Tboefijn — APPROVED** [review]
LGTM!

File: Aplib.Core/Intent/Actions/QueryAction.cs [hunk 1]
```diff
@@ -25,30 +25,39 @@ public class GuardedAction<TBeliefSet, TQuery> : Action<TBeliefSet>
...
-        public override bool IsActionable(TBeliefSet beliefSet)
+        /// <summary>
+        /// Queries the environment for the guarded item and returns whether the guard is not null.
+        /// </summary>
+        /// <param name="beliefSet">The belief set of the agent.</param>
+        /// <returns>True if the guard is not null; otherwise, false.</returns>
+        public bool Query(TBeliefSet beliefSet)
         {
-            _storedGuardResult = _guard(beliefSet);
+            // Query the environment for the guarded item
```
Comment (JensSteenmetz) [inline Aplib.Core/Intent/Actions/QueryAction.cs]: Nit
```suggestion
            // Query the environment for the guarded item.
```

File: Aplib.Core/Intent/Actions/QueryAction.cs [hunk 2]
```diff
@@ -25,30 +25,39 @@ public class GuardedAction<TBeliefSet, TQuery> : Action<TBeliefSet>
...
+            _storedGuardResult = _query(beliefSet);
+
+            // Only return true if the guard is not null
```
Comment (JensSteenmetz) [inline Aplib.Core/Intent/Actions/QueryAction.cs]: Nit
```suggestion
            // Only return true if the guard is not null.
```

File: Aplib.Core/Intent/Actions/QueryAction.cs [hunk 3]
```diff
@@ -25,30 +25,39 @@ public class GuardedAction<TBeliefSet, TQuery> : Action<TBeliefSet>
...
+        /// <summary>
+        /// Queries the environment for the guarded item and returns whether the guard is not null.
+        /// </summary>
+        /// <param name="beliefSet">The belief set of the agent.</param>
+        /// <returns>True if the guard is not null; otherwise, false.</returns>
+        public bool Query(TBeliefSet beliefSet)
         {
-            _storedGuardResult = _guard(beliefSet);
+            // Query the environment for the guarded item
+            _storedGuardResult = _query(beliefSet);
```
Comment (JensSteenmetz) [inline Aplib.Core/Intent/Actions/QueryAction.cs]:
```suggestion
            _storedQueryResult = _query(beliefSet);
```

File: Aplib.Core/Intent/Actions/QueryAction.cs [hunk 4]
```diff
@@ -25,30 +25,39 @@ public class GuardedAction<TBeliefSet, TQuery> : Action<TBeliefSet>
...
-        protected new System.Func<TBeliefSet, TQuery?> _guard { get; set; }
+        protected System.Func<TBeliefSet, TQuery?> _query { get; set; }
 
         /// <summary>
-        /// Initializes a new instance of the <see cref="GuardedAction{TBeliefSet,TQuery}"/> class.
+        /// Initializes a new instance of the <see cref="QueryAction{TBeliefSet,TQuery}"/> class.
         /// </summary>
         /// <param name="effect">The effect of the action.</param>
-        /// <param name="guard">The guard of the action.</param>
+        /// <param name="query">The guard of the action.</param>
```
Comment (JensSteenmetz) [inline Aplib.Core/Intent/Actions/QueryAction.cs]: There are more comments that need to be addressed.
```suggestion
        /// <param name="query">The query of the action.</param>
```

File: Aplib.Core/Intent/Tactics/PrimitiveTactic.cs [hunk 5]
```diff
@@ -33,13 +33,32 @@ public PrimitiveTactic(IAction<TBeliefSet> action, Metadata? metadata = null)
...
+        /// </summary>
+        /// <param name="queryAction">The queryable action of the primitive tactic.</param>
+        /// <param name="guard">The guard of the primitive tactic.</param>
+        /// <param name="metadata">
+        /// Metadata about this tactic, used to quickly display the tactic in several contexts.
+        /// </param>
+        public PrimitiveTactic(IQueryable<TBeliefSet> queryAction,
+            System.Func<TBeliefSet, bool> guard,
+            Metadata? metadata = null)
+            : base((beliefset) => guard(beliefset) && queryAction.Query(beliefset))
```
Comment (JensSteenmetz) [inline Aplib.Core/Intent/Tactics/PrimitiveTactic.cs]:
```suggestion
            : base(beliefSet => guard(beliefSet) && queryAction.Query(beliefSet), metadata)
```

File: Aplib.Core/Intent/Tactics/PrimitiveTactic.cs [hunk 6]
```diff
@@ -33,13 +33,32 @@ public PrimitiveTactic(IAction<TBeliefSet> action, Metadata? metadata = null)
...
+        {
+            _action = queryAction;
+        }
```
Comment (JensSteenmetz) [inline Aplib.Core/Intent/Tactics/PrimitiveTactic.cs]: Nit
```suggestion
        =>  _action = queryAction;
```

**Review by JensSteenmetz — CHANGES_REQUESTED** [review]
Great work, I commented on some details

File: Aplib.Core/Intent/Actions/QueryAction.cs [hunk 7]
```diff
@@ -1,10 +1,10 @@
 ﻿using Aplib.Core.Belief;
-using Aplib.Core.Intent.Actions;
+using System;
```
Comment (JensSteenmetz) [inline Aplib.Core/Intent/Actions/QueryAction.cs]: Could you remove using System? It sometimes creates issues with our Action class vs the System.Action delegate.

File: Aplib.Core/Intent/Actions/QueryAction.cs [hunk 8]
```diff
@@ -13,52 +13,52 @@ public class QueryAction<TBeliefSet, TQuery> : Action<TBeliefSet>, IQueryable<TB
...
             : base(metadata)
         {
             _effect = effect;
             _query = query;
         }
 
-        /// <inheritdoc/>
-        public override void Execute(TBeliefSet beliefSet) => _effect(beliefSet, _storedGuardResult!);
+        /// <inheritdoc />
+        public override void Execute(TBeliefSet beliefSet) => _effect(beliefSet, arg2: _storedQueryResult!);
```
Comment (JensSteenmetz) [inline Aplib.Core/Intent/Actions/QueryAction.cs]:
```suggestion
        public override void Execute(TBeliefSet beliefSet) => _effect(beliefSet, _storedQueryResult!);
```

**Review by JensSteenmetz — CHANGES_REQUESTED** [review]
Almost ;)

**Review by JensSteenmetz — APPROVED** [review]
LGTM!!!

**Review by Tboefijn — APPROVED** [review]
LGTM!

**Pull request merged**